mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,279 @@
+
+        for (size_t i = 0; i < OpsEachTape.size(); i++) {
+            OpsEachTape[i] = std::make_shared<SmallVector<Operation *>>();
+        }
+
+        // special: [0] for pre and [-1] for post processing
+        unsigned int cur_tape = 0;
+
+        func->walk([&](Operation *op) {
+            if (op == func) {
+                return; // don't visit the funcop itself
+            }
+            if (op->getName().getStringRef() == "quantum.device") {
+                cur_tape++;
+            }
+            OpsEachTape[cur_tape]->push_back(op);
+            if ((op->getName().getStringRef() == "quantum.device_release") &&
+                cur_tape == OpsEachTape.size() - 2) {
+                // reached post processing
+                cur_tape++;
+            }
+        });
+    }
+
+    LogicalResult CreateTapeFunction(std::shared_ptr<SmallVector<Operation *>> TapeOps,
+                                     SmallVector<Value> &NecessaryValuesFromEarlierTapes,
+                                     IRRewriter &builder, unsigned int tape_number,
+                                     func::FuncOp OriginalMultitapeFunc,
+                                     SmallVector<FailureOr<func::FuncOp>> &OutlinedFuncs)
+    {
+        assert(TapeOps); // nullptr protection
+
+        // 1. Identify the necessary return values
+        SmallVector<Value> RetValues;
+        SmallVector<mlir::Type> RetTypes;
+
+        for (Value v : NecessaryValuesFromEarlierTapes) {
+            Operation *VDefOp = v.getDefiningOp();
+            if (std::find(TapeOps->begin(), TapeOps->end(), VDefOp) != TapeOps->end()) {
+                // This Value needed for PP is in this tape!
+                RetValues.push_back(v);
+                RetTypes.push_back(v.getType());
+            }
+        }
+
+        // We process the tapes in reverse order
+        // Hence from the viewpoint of the earlier tapes,
+        // a processed later tape (which has become a call op) is also their "post processing"
+        // The later tape's arguments need to be returned by the earlier tapes
+        // Hence when processing a tape, it needs to ask for earlier tapes to return
+        // values it needs, just like how post processing asks for all tapes to return
+        // values it needs.
+        for (Operation *op : *TapeOps) {
+            for (auto operand : op->getOperands()) {
+                Operation *OperandSource = operand.getDefiningOp();
+                if (std::find(TapeOps->begin(), TapeOps->end(), OperandSource) == TapeOps->end()) {
+                    // A tape operand not produced in this tape itself,
+                    // must be from the tapes/preprocessing!
+                    // Need to be replaced by the previous tape functions' return values
+                    NecessaryValuesFromEarlierTapes.push_back(operand);
+                }
+            }
+        }
+
+        // 2. Create a scf.execute_region and wrap it around the tape ops
```

Comment:
Can you outline this into its own function?

Reply:
done